x/appchain/coordinator/keeper/slash.go

```diff
@@ -3,12 +3,101 @@ package keeper
 import (
 	"time"
 
+	"github.com/ExocoreNetwork/exocore/utils"
+	commontypes "github.com/ExocoreNetwork/exocore/x/appchain/common/types"
 	types "github.com/ExocoreNetwork/exocore/x/appchain/coordinator/types"
 	sdk "github.com/cosmos/cosmos-sdk/types"
 )
 
-// TODO: this file should be in the x/avs keeper instead.
+// ValidateSlashPacket validates a slashing packet. It checks that
+// (1) the valset update id maps back to an Exocore height, and
+// (2) the validator (cons) address maps back to an operator account address.
+// The caller must perform stateless validation by themselves.
+func (k Keeper) ValidateSlashPacket(
+	ctx sdk.Context, chainID string, data commontypes.SlashPacketData,
+) error {
+	// the validator set is generated at each epoch and following each slash. even if the set
+	// does not change at each epoch, we send an empty update anyway since, otherwise, the
+	// subscriber could time out the coordinator.
+	if k.GetHeightForChainVscID(ctx, chainID, data.ValsetUpdateID) == 0 {
+		return commontypes.ErrInvalidPacketData.Wrapf(
+			"invalid chainID %s valsetUpdateID %d", chainID, data.ValsetUpdateID,
+		)
+	}
+	// the second step is to find the operator account address against the consensus address
+	// of the validator. if the operator is not found, the slashing packet is invalid.
+	// note that this works even if the operator changes their consensus key, as long as the
+	// key hasn't yet been pruned from the operator module.
+	if found, _ := k.operatorKeeper.GetOperatorAddressForChainIDAndConsAddr(
+		ctx, chainID, sdk.ConsAddress(data.Validator.Address),
+	); !found {
+		// don't bech32 encode it in the error since the appchain may have a different prefix
+		return commontypes.ErrInvalidPacketData.Wrapf(
+			"operator not found %x", data.Validator.Address,
+		)
+	}
+	return nil
+}
+
+// HandleSlashPacket handles a slashing packet. The caller must ensure that the slashing packet
+// is valid before calling this function. The function forwards the slashing request to the
+// operator module, which will trigger a slashing hook and thus a validator set update containing
+// the slashing acknowledgment.
+func (k Keeper) HandleSlashPacket(ctx sdk.Context, chainID string, data commontypes.SlashPacketData) {
+	consAddress := sdk.ConsAddress(data.Validator.Address)
+	// never 0, since already validated
+	height := k.GetHeightForChainVscID(ctx, chainID, data.ValsetUpdateID)
```

Comment:
No. It could technically be mapped using the epoch identifier and number; however, mid-epoch slashing would need a separate identifier and number.

Reply:
I changed the design to use epoch identifier and number as the snapshot key when implementing it. The main reason is that with the introduction of `VotingPowerActivationDuration`. When `VotingPowerActivationDuration` is greater than 1, it becomes difficult to retrieve the `start_height` of future epochs when saving snapshots. However, the benefit of incorporating epoch information into the key is that, if both epoch and height information are provided, the corresponding snapshot can be quickly retrieved. So they are expected to be provided when addressing slash. But I'm also not very sure whether it's fine for this implementation.

Reply:
I will change the key of the voting power snapshot to height and deprecate `VotingPowerActivationDuration`, because only the height can be provided when addressing the slash from CometBFT.